tunnel/ssh.go

```diff
@@ -0,0 +1,161 @@
+package tunnel
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+	"sync"
+	"time"
+
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/sync/errgroup"
+)
+
+type SSHConfig struct {
+	User       string
+	Host       string
+	Port       int
+	PrivateKey []byte
+
+	RemoteHost string
+	RemotePort int
+}
+
+type SSH struct {
+	localServer net.Listener
+	sshClient   *ssh.Client
+	remoteAddr  string
+
+	err   error
+	errMu sync.Mutex
+
+	backgroundWG sync.WaitGroup
+}
+
+func Listen(config *SSHConfig) (*SSH, error) {
+	singer, err := ssh.ParsePrivateKey(config.PrivateKey)
+	if err != nil {
+		return nil, fmt.Errorf("parsing private key: %s", err.Error())
+	}
+
+	endpoint := net.JoinHostPort(config.Host, strconv.Itoa(config.Port))
+	sshClient, err := ssh.Dial("tcp", endpoint, &ssh.ClientConfig{
+		User: config.User,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(singer),
+		},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+		BannerCallback:  ssh.BannerDisplayStderr(),
+		Timeout:         10 * time.Second,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("server %q dial error: %w", endpoint, err)
+	}
+
+	listener, err := net.Listen("tcp", `127.0.0.1:0`)
+	if err != nil {
+		sshClient.Close()
+		return nil, fmt.Errorf("listening on %s://%s: %w", "tcp", listener.Addr().String(), err)
+	}
+
+	tunnel := &SSH{
+		localServer: listener,
+		sshClient:   sshClient,
+		remoteAddr: net.JoinHostPort(
+			config.RemoteHost,
+			strconv.Itoa(config.RemotePort),
+		),
+	}
+
+	go tunnel.listen()
```

Comment:
I believe not because we have only started listening for connections.

If I remember correctly, there is a buffering between the connection being established and accept is called to handle it.

I keep a not to research and confirm this, as well as adding stress tests around this.